File: iter1/39.cpp
// 思路：dfs，每一层选一个数字。注意要点是，下一层不能选前一层之前的数字。


#include <vector>
#include <iostream>

using namespace std;

class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        return combinationSum(candidates, target, 0);
    }

    vector<vector<int>> combinationSum(vector<int>& candidates, int target, int left) {
        vector<vector<int>> ret;
        for (int i = left; i < candidates.size(); ++i) {
            if (target - candidates[i] < 0) continue;
            else if (target - candidates[i] == 0) ret.push_back({candidates[i]});
            else {
                auto nextRet = combinationSum(candidates, target - candidates[i], i);
                for (auto &row : nextRet) {
                    row.push_back(candidates[i]);
                    ret.push_back(row);
                }
            }
        }
        return ret;
    }
};


int main() {
    Solution s;
    vector<int> candidates{2,3,5};
    auto res = s.combinationSum(candidates, 8);
    for (auto &row : res) {
        for (auto n : row) {
            cout << n << " ";
        }
        cout << endl;
    }
}

File: iter2/39.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        ans = []
        def dfs(remain, nums, i):
            if remain == 0:
                ans.append(nums)
                return
            if remain < 0: return
            if i >= len(candidates): return
            for q in range(0, (candidates[i] // remain)+1):
                dfs(remain-candidates[i]*q, nums+[candidates[i]]*q, i+1)
        dfs(target, [], 0)
        return ans


class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        candidates.sort()
        ans = []
        # pick zero or unlimited of candidates[i]
        def dfs(s, used, remain):
            for i in range(s, len(candidates)):
                c = candidates[i]
                if c == remain:
                    ans.append(used + [c])
                elif c < remain:
                    dfs(i, used+[c], remain-c)
                elif c > remain:
                    return
        dfs(0, [], target)
        # print(ans)
        return ans





